cmd/restic/cmd_backup.go

```diff
@@ -522,3 +544,87 @@ func readExcludePatternsFromFiles(excludeFiles []string) []string {
 	}
 	return excludes
 }
+
+// FilenameCheck is a function that takes a filename and returns a boolean depending on arbitraty check
+type FilenameCheck func(filename string) bool
+
+// genExcludeByFile returns a FilenameCheck which itself returns whether a path
+// should be excluded. The FilenameCheck considers a file to be excluded when
+// it resides in a directory with an exclusion file, that is either specified
+// by excludeIfPresent in the form "filename[:content]".  The returned error is
+// non-nil if the filename component of excludeIfPresent is empty.
+func genExcludeByFile(excludeIfPresent string) (FilenameCheck, error) {
+	fn := func(string) bool { return false }
+	tf, tc := "", ""
+	if excludeIfPresent != "" {
+		colon := strings.Index(excludeIfPresent, ":")
+		switch {
+		default:
+			tf = excludeIfPresent[:colon]
+			tc = excludeIfPresent[colon+1:]
+		case colon < 0:
+			tf = excludeIfPresent
+		case colon == 0:
+			return nil, fmt.Errorf("no name for exclusion tagfile provided")
+		}
+	}
+	if tf != "" {
+		debug.Log("using %q as exclusion tagfile", tf)
+		fn = func(filename string) bool {
+			return isExcludedByFile(filename, tf, tc)
+		}
+	}
+	return fn, nil
+}
+
+// isExcludedByFile interprets filename as a path and returns true if that file
+// is in a excluded directory. A directory is identified as excluded if it contains a
+// tagfile which bears the name specified in tagFilename and starts with header.
+func isExcludedByFile(filename, tagFilename, header string) bool {
+	if tagFilename == "" {
+		return false
+	}
+	dir, base := filepath.Split(filename)
+	if base == tagFilename {
+		return false // do not exclude the tagfile itself
+	}
+	tf := filepath.Join(dir, tagFilename)
+	_, err := os.Stat(tf)
```

Comment:
Use `github.com/restic/restic/internal/fs` instead of `os`, for Windows that's important to correctly handle long paths.

In addition: I think it is necessary to use `Lstat()` here: If the tag file is a symlink, `Stat()` returns an error if the target does not exist. What do you think?

Reply:
Thanks for the heads-up regarding `internal/fs`, I wasn't aware of that. Done.
`Lstat` also done.